server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashType.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.geogrid;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Locale;
+
+public enum GeoHashType implements Writeable {
+    GEOHASH(new GeoHashHandler());
+
+    public static final GeoHashType DEFAULT = GEOHASH;
+
+    private GeoHashTypeProvider handler;
+
+    GeoHashType(GeoHashTypeProvider handler) {
+        this.handler = handler;
+    }
+
+    /**
+     * Case-insensitive from string method.
+     *
+     * @param value String representation
+     * @return The hash type
+     */
+    public static GeoHashType forString(String value) {
+        return valueOf(value.toUpperCase(Locale.ROOT));
+    }
+
+    public static GeoHashType readFromStream(StreamInput in) throws IOException {
+
+        // FIXME: update version after backport
+
+        if(in.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+            return in.readEnum(GeoHashType.class);
+        } else {
+            return DEFAULT;
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+
+        // FIXME: update version after backport
+
+        // To maintain binary compatibility, only include type after the given version
+        if (out.getVersion().onOrAfter(Version.V_7_0_0_alpha1)) {
+            out.writeEnum(this);
+        } else if (this != DEFAULT) {
+            throw new UnsupportedOperationException("Geo aggregation type [" + this +
+                "] is not supported by the node version " + out.getVersion().toString());
+        }
+    }
```

Comment:
We should write a test for this Enum. Have a look at ShapeRelationTests to see the kind of tests I mean. This will ensure that if the ordering of the enum is changed or if a new enum value is added it will be catch any changes to serialisation. We should also add a test for reading and writing the enum to and from a version before this feature

Reply:
thx, done - see last commit.